test/framework/src/main/java/org/elasticsearch/transport/nio/channel/TcpReadContext.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.transport.nio.channel;
+
+import org.elasticsearch.common.bytes.BytesArray;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.transport.nio.ByteBufferReference;
+import org.elasticsearch.transport.nio.CompositeByteBufferReference;
+import org.elasticsearch.transport.nio.TcpReadHandler;
+
+import java.io.IOException;
+
+public class TcpReadContext implements ReadContext {
+
+    private static final int DEFAULT_READ_LENGTH = 1 << 14;
+
+    private final TcpReadHandler handler;
+    private final NioSocketChannel channel;
+    private final TcpFrameDecoder frameDecoder;
+    private final CompositeByteBufferReference references;
+
+    public TcpReadContext(NioSocketChannel channel, TcpReadHandler handler) {
+        this(channel, handler, new TcpFrameDecoder());
+    }
+
+    public TcpReadContext(NioSocketChannel channel, TcpReadHandler handler, TcpFrameDecoder frameDecoder) {
+        this.handler = handler;
+        this.channel = channel;
+        this.frameDecoder = frameDecoder;
+        this.references = new CompositeByteBufferReference(ByteBufferReference.heapBuffer(new BytesArray(new byte[DEFAULT_READ_LENGTH])));
```

Comment:
I'd just don't worry about all the default read length and use `PagedBytesReference` and big arrays here. We can also do this in a followup and we just don't do any reuse etc. and simply / blindly always create a fresh byte array. also look at my suggestion above in `NioSocketChannel`